src/pybreaker.py

```diff
@@ -606,6 +606,159 @@ def _namespace(self, key):
+
+        self._initialize_redis_state(self._initial_state)
+
+    def _initialize_redis_state(self, state):
+        self._redis.setnx(self._namespace('fail_counter'), 0)
+        self._redis.setnx(self._namespace('state'), state)
+
+    @property
+    def state(self):
+        """
+        Returns the current circuit breaker state.
+
+        If the circuit breaker state on Redis is missing, re-initialize it
+        with the fallback circuit state and reset the fail counter.
+        """
+        try:
+            state_bytes = self._redis.get(self._namespace('state'))
+        except self.RedisError:
+            self.logger.error('RedisError: falling back to default circuit state', exc_info=True)
+            return self._fallback_circuit_state
+
+        state = self._fallback_circuit_state
+        if state_bytes is not None:
+            state = state_bytes.decode('utf-8')
+        else:
+            # state retrieved from redis was missing, so we re-initialize
+            # the circuit breaker state on redis
+            self._initialize_redis_state(self._fallback_circuit_state)
+
+        return state
+
+    @state.setter
+    def state(self, state):
+        """
+        Set the current circuit breaker state to `state`.
+        """
+        try:
+            self._redis.set(self._namespace('state'), str(state))
+        except self.RedisError:
+            self.logger.error('RedisError', exc_info=True)
+            pass
+
+    def increment_counter(self):
+        """
+        Increases the failure counter by one.
+        """
+        try:
+            self._redis.incr(self._namespace('fail_counter'))
+        except self.RedisError:
+            self.logger.error('RedisError', exc_info=True)
+            pass
+
+    def reset_counter(self):
+        """
+        Sets the failure counter to zero.
+        """
+        try:
+            self._redis.set(self._namespace('fail_counter'), 0)
+        except self.RedisError:
+            self.logger.error('RedisError', exc_info=True)
+            pass
+
+    @property
+    def counter(self):
+        """
+        Returns the current value of the failure counter.
+        """
+        try:
+            value = self._redis.get(self._namespace('fail_counter'))
+            if value:
+                return int(value)
+            else:
+                return 0
+        except self.RedisError:
+            self.logger.error('RedisError: Assuming no errors', exc_info=True)
+            return 0
+
+    @property
+    def opened_at(self):
+        """
+        Returns a datetime object of the most recent value of when the circuit
+        was opened.
+        """
+        try:
+            timestamp = self._redis.get(self._namespace('opened_at'))
+            if timestamp:
+                return datetime(*time.gmtime(int(timestamp))[:6])
+        except self.RedisError:
+            self.logger.error('RedisError', exc_info=True)
+            return None
```

Comment:
As you said, this is the only bit that changes from the current implementation.

Can't we find a way to add support for this without having to duplicate the whole class?

Reply:
yes, I agree, I adjusted it by creating an extra parameter.